ul/src/pdu/writer.rs

```diff
@@ -0,0 +1,745 @@
+        writer.write_u8(presentation_context.id)?;
+
+        // 6 - Reserved - This reserved field shall be sent with a value 00H but not tested to this
+        // value when received.
+        writer.write_u8(0x00)?;
+
+        // 7 - Result/Reason - This Result/Reason field shall contain an integer value encoded as an
+        // unsigned binary number. One of the following values shall be used:
+        //   0 - acceptance
+        //   1 - user-rejection
+        //   2 - no-reason (provider rejection)
+        //   3 - abstract-syntax-not-supported (provider rejection)
+        //   4 - transfer-syntaxes-not-supported (provider rejection)
+        match &presentation_context.reason {
+            PresentationContextResultReason::Acceptance => {
+                writer.write_u8(0)?;
+            }
+            PresentationContextResultReason::UserRejection => {
+                writer.write_u8(1)?;
+            }
+            PresentationContextResultReason::NoReason => {
+                writer.write_u8(2)?;
+            }
+            PresentationContextResultReason::AbstractSyntaxNotSupported => {
+                writer.write_u8(3)?;
+            }
+            PresentationContextResultReason::TransferSyntaxesNotSupported => {
+                writer.write_u8(4)?;
+            }
+        }
+
+        // 8 - Reserved - This reserved field shall be sent with a value 00H but not tested to this
+        // value when received.
+        writer.write_u8(0x00)?;
+
+        // 9-xxx - Transfer syntax sub-item - This variable field shall contain one Transfer Syntax
+        // Sub-Item. When the Result/Reason field has a value other than acceptance (0), this field
+        // shall not be significant and its value shall not be tested when received. For a complete
+        // description of the use and encoding of this item see Section 9.3.3.2.1.
+
+        // 1 - Item-type - 40H
+        writer.write_u8(0x40)?;
+
+        // 2 - Reserved - This reserved field shall be sent with a value 00H but not tested to this
+        // value when received.
+        writer.write_u8(0x40)?;
+
+        write_chunk_u16(writer, |writer| {
+            // 5-xxx - Transfer-syntax-name - This variable field shall contain the
+            // Transfer-syntax-name proposed for this presentation context. A valid
+            // Transfer-syntax-name shall be encoded as defined in Annex F. For a description of the
+            // use of this field see Section 7.1.1.14. Transfer-syntax-names are structured as UIDs
+            // as defined in PS3.5 (see Annex B for an overview of this concept). DICOM
+            // Transfer-syntax-names are registered in PS3.5.
+            write_n(writer, presentation_context.transfer_syntax.as_bytes())?;
+
+            Ok(())
+        })?;
+
+        Ok(())
+    })?;
+
+    Ok(())
+}
+
+fn write_pdu_variable_user_variables(
+    writer: &mut dyn Write,
+    user_variables: &Vec<UserVariableItem>,
```

Comment:
`user_variables` can be of type `&[UserVariableItem]` instead. SO question with greater detail: https://stackoverflow.com/q/40006219/1233251